### PR TITLE
add IKP_Translation{X,Y,Z}AxisAngle{Z,X,Y}Norm4D to the cpp template

### PR DESCRIPTION
**Pull request by k-okada**
based on comment https://github.com/ros-planning/moveit/issues/548#issuecomment-316298918 and I have tested on https://github.com/k-okada/SCORE/tree/master/src/scara_ikfast_plugin

### CODE REVIEW

File: moveit_kinematics/ikfast_kinematics_plugin/templates/ikfast61_moveit_plugin_template.cpp [hunk 1]
```diff
@@ -544,12 +544,20 @@ int IKFastKinematicsPlugin::solve(KDL::Frame& pose_frame, const std::vector<doub
       return solutions.GetNumSolutions();
 
     case IKP_TranslationXAxisAngle4D:
```
Comment (mintar) [inline moveit_kinematics/ikfast_kinematics_plugin/templates/ikfast61_moveit_plugin_template.cpp]: Please add a comment from the latest ikfast documentation, found here: http://openrave.org/docs/latest_stable/openravepy/ikfast/

> * TranslationXAxisAngle4D, TranslationYAxisAngle4D, TranslationZAxisAngle4D - end effector origin reaches desired 3D translation, manipulator direction makes a specific angle with x/y/z-axis (defined in the manipulator base link’s coordinate system)

File: moveit_kinematics/ikfast_kinematics_plugin/templates/ikfast61_moveit_plugin_template.cpp [hunk 2]
```diff
@@ -544,12 +544,20 @@ int IKFastKinematicsPlugin::solve(KDL::Frame& pose_frame, const std::vector<doub
...
+      double roll, pitch, yaw;
+      pose_frame.M.GetRPY(roll, pitch, yaw);
```
Comment (mintar) [inline moveit_kinematics/ikfast_kinematics_plugin/templates/ikfast61_moveit_plugin_template.cpp]: According to the documentation snippet I've pasted above, and to https://github.com/rdiankov/openrave/issues/596#issuecomment-403673192, I think just using `GetRPY()` is wrong for `Translation*AxisAngle4D`. The specific angle you pass into `ComputeIk` (here: `&roll`) should be the angle between the manipulator direction and the X axis of the  manipulator base link’s coordinate system. Same for `TranslationYAxisAngle4D` and `TranslationZAxisAngle4D`.

File: moveit_kinematics/ikfast_kinematics_plugin/templates/ikfast61_moveit_plugin_template.cpp [hunk 3]
```diff
@@ -561,11 +569,23 @@ int IKFastKinematicsPlugin::solve(KDL::Frame& pose_frame, const std::vector<doub
     case IKP_Lookat3D:
     case IKP_TranslationXY2D:
     case IKP_TranslationXYOrientation3D:
+      ROS_ERROR_NAMED(name_, "IK for this IkParameterizationType not implemented yet.");
+      return 0;
+
     case IKP_TranslationXAxisAngleZNorm4D:
```
Comment (mintar) [inline moveit_kinematics/ikfast_kinematics_plugin/templates/ikfast61_moveit_plugin_template.cpp]: Please add a comment from the latest ikfast documentation:

> * TranslationXAxisAngleZNorm4D - end effector origin reaches desired 3D translation, manipulator direction needs to be orthogonal to z axis and be rotated at a certain angle starting from the x axis (defined in the manipulator base link’s coordinate system)

File: moveit_kinematics/ikfast_kinematics_plugin/templates/ikfast61_moveit_plugin_template.cpp [hunk 4]
```diff
@@ -561,11 +569,23 @@ int IKFastKinematicsPlugin::solve(KDL::Frame& pose_frame, const std::vector<doub
...
+      pose_frame.M.GetRPY(roll, pitch, yaw);
+      ComputeIk(trans, &yaw, vfree.size() > 0 ? &vfree[0] : NULL, solutions);
+      return solutions.GetNumSolutions();
+
     case IKP_TranslationYAxisAngleXNorm4D:
```
Comment (mintar) [inline moveit_kinematics/ikfast_kinematics_plugin/templates/ikfast61_moveit_plugin_template.cpp]: Please add a comment from the latest ikfast documentation:

> * TranslationYAxisAngleXNorm4D - end effector origin reaches desired 3D translation, manipulator direction needs to be orthogonal to x axis and be rotated at a certain angle starting from the y axis (defined in the manipulator base link’s coordinate system)

File: moveit_kinematics/ikfast_kinematics_plugin/templates/ikfast61_moveit_plugin_template.cpp [hunk 5]
```diff
@@ -561,11 +569,23 @@ int IKFastKinematicsPlugin::solve(KDL::Frame& pose_frame, const std::vector<doub
...
+      pose_frame.M.GetRPY(roll, pitch, yaw);
+      ComputeIk(trans, &roll, vfree.size() > 0 ? &vfree[0] : NULL, solutions);
+      return solutions.GetNumSolutions();
+
     case IKP_TranslationZAxisAngleYNorm4D:
```
Comment (mintar) [inline moveit_kinematics/ikfast_kinematics_plugin/templates/ikfast61_moveit_plugin_template.cpp]: Please add a comment from the latest ikfast documentation:

> * TranslationZAxisAngleYNorm4D - end effector origin reaches desired 3D translation, manipulator direction needs to be orthogonal to y axis and be rotated at a certain angle starting from the z axis (defined in the manipulator base link’s coordinate system)

File: moveit_kinematics/ikfast_kinematics_plugin/templates/ikfast61_moveit_plugin_template.cpp [hunk 6]
```diff
@@ -561,11 +569,23 @@ int IKFastKinematicsPlugin::solve(KDL::Frame& pose_frame, const std::vector<doub
...
-      ROS_ERROR_NAMED(name_, "IK for this IkParameterizationType not implemented yet.");
-      return 0;
+      pose_frame.M.GetRPY(roll, pitch, yaw);
+      ComputeIk(trans, &pitch, vfree.size() > 0 ? &vfree[0] : NULL, solutions);
+      return solutions.GetNumSolutions();
 
     default:
       ROS_ERROR_NAMED(name_, "Unknown IkParameterizationType! Was the solver generated with an incompatible version "
```
Comment (mintar) [inline moveit_kinematics/ikfast_kinematics_plugin/templates/ikfast61_moveit_plugin_template.cpp]: General comment: Could you modify your code again and verify that it works with this PR + `TranslationXAxisAngleZNorm4D`, as suggested in https://github.com/rdiankov/openrave/issues/596#issuecomment-403673192 ?

File: moveit_kinematics/ikfast_kinematics_plugin/templates/ikfast61_moveit_plugin_template.cpp (hunk 1 above)
Comment (mintar) [reply moveit_kinematics/ikfast_kinematics_plugin/templates/ikfast61_moveit_plugin_template.cpp]: Done.

File: moveit_kinematics/ikfast_kinematics_plugin/templates/ikfast61_moveit_plugin_template.cpp (hunk 3 above)
Comment (mintar) [reply moveit_kinematics/ikfast_kinematics_plugin/templates/ikfast61_moveit_plugin_template.cpp]: Done.

File: moveit_kinematics/ikfast_kinematics_plugin/templates/ikfast61_moveit_plugin_template.cpp (hunk 4 above)
Comment (mintar) [reply moveit_kinematics/ikfast_kinematics_plugin/templates/ikfast61_moveit_plugin_template.cpp]: Done.

File: moveit_kinematics/ikfast_kinematics_plugin/templates/ikfast61_moveit_plugin_template.cpp (hunk 6 above)
Comment (k-okada) [reply moveit_kinematics/ikfast_kinematics_plugin/templates/ikfast61_moveit_plugin_template.cpp]: Thank you for comment, Actually I do not have robot model to test `Translation$AxisAngle4D` so I think this PR should only deal with 'Translation$AxisAngle"#orm4D` and work for `Translation$AxisAngle4D`  when we found URDF to test.
I have confirmed with this new PR, see https://github.com/k-okada/SCORE/tree/master/src/scara_ikfast_plugin and https://github.com/ros-planning/moveit/issues/548#issuecomment-403674036

File: moveit_kinematics/ikfast_kinematics_plugin/templates/ikfast61_moveit_plugin_template.cpp (hunk 5 above)
Comment (mintar) [reply moveit_kinematics/ikfast_kinematics_plugin/templates/ikfast61_moveit_plugin_template.cpp]: Done.

File: moveit_kinematics/ikfast_kinematics_plugin/templates/ikfast61_moveit_plugin_template.cpp (hunk 6 above)
Comment (mintar) [reply moveit_kinematics/ikfast_kinematics_plugin/templates/ikfast61_moveit_plugin_template.cpp]: I agree.

**Comment by mintar** [comment]
LGTM. No idea why the CI checks are failing now, this shouldn't be related to the latest commit.

**Comment by k-okada** [comment]
```
2.11s$ git clone -q https://github.com/ros-planning/moveit_ci.git .moveit_ci
82.29s$ .moveit_ci/travis.sh
---
Testing branch 'kinetic-devel' of 'moveit' on ROS 'kinetic'
Starting Docker image: moveit/moveit:kinetic-ci
docker: invalid reference format.
See 'docker run --help'.
.moveit_ci/travis.sh: line 56: moveit/moveit:kinetic-ci: No such file or directory
moveit/moveit:kinetic-ci container finished with errors
```
 should I target to melodic?

**Comment by k-okada** [comment]
ok, finally passed the travis.

**Pull request merged**